src/Microsoft.ApplicationInsights.AspNetCore/TelemetryInitializers/AzureAppServiceRoleNameFromHostNameHeaderInitializer.cs

```diff
@@ -0,0 +1,149 @@
+        private readonly IHttpContextAccessor httpContextAccessor;
+        private string roleName;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AzureAppServiceRoleNameFromHostNameHeaderInitializer" /> class.
+        /// </summary>
+        /// <param name="httpContextAccessor">Accessor to provide HttpContext if available.</param>
+        public AzureAppServiceRoleNameFromHostNameHeaderInitializer(IHttpContextAccessor httpContextAccessor)
+        {
+            this.httpContextAccessor = httpContextAccessor ?? throw new ArgumentNullException(nameof(httpContextAccessor));
+
+            try
+            {
+                var result = Environment.GetEnvironmentVariable(WebAppHostNameEnvironmentVariable);
+
+                if (!string.IsNullOrEmpty(result) && result.EndsWith(this.WebAppSuffix, StringComparison.OrdinalIgnoreCase))
+                {
+                    result = result.Substring(0, result.Length - this.WebAppSuffix.Length);
+                }
+
+                this.roleName = result;
+            }
+            catch (Exception ex)
+            {
+                AspNetCoreEventSource.Instance.LogAzureAppServiceRoleNameFromHostNameHeaderInitializerWarning(ex.ToInvariantString());
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets suffix of website name. This must be changed when running in non public Azure region.
+        /// Default value (Public Cloud):  ".azurewebsites.net"
+        /// For US Gov Cloud: ".azurewebsites.us"
+        /// For Azure Germany: ".azurewebsites.de".
+        /// </summary>
+        public string WebAppSuffix { get; set; } = ".azurewebsites.net";
+
+        /// <summary>
+        /// Populates RoleName from the request telemetry associated with the http context.
+        /// If RoleName is empty on the request telemetry, it'll be updated as well so that other telemetry
+        /// belonging to the same requests gets it from request telemetry, without having to parse headers again.
+        /// </summary>
+        /// <remarks>
+        /// RoleName is attempted from every incoming request as opposed to doing this periodically. This is
+        /// done to ensure every request (and associated telemetry) gets the correct RoleName during slot swap.
+        /// </remarks>
+        /// <param name="telemetry">The telemetry item for which RoleName is to be set.</param>
+        public void Initialize(ITelemetry telemetry)
+        {
+            try
+            {
+                if (!string.IsNullOrEmpty(telemetry.Context.Cloud.RoleName))
+                {
+                    // RoleName is already populated.
+                    return;
+                }
+
+                string roleName = string.Empty;
+                var context = this.httpContextAccessor.HttpContext;
+
+                if (context != null)
+                {
+                    lock (context)
```

Comment:
can you add a comment.. I don't understand why is the context being locked here?

Reply:
One of the oldest issues we have: https://github.com/microsoft/ApplicationInsights-aspnetcore/issues/373
(if you check all other initializers in this repo, you'll see same lock)

Reply:
you could easily avoid lock and access to the http context if you parsed the header in the request tracking module and populated it somewhere. Initializer would took the value from there.
This avoids accessing HttpContext on each telemetry and lock on each telemetry.

Please also consider side effects like one we learned yesterday about (https://github.com/aspnet/AspNetCore/issues/14975) that availability and integrity of HttpContext at the moment when telemetry is tracked is not guaranteed and overall, HttpContext is not recommended API to use.

Reply:
Yes its some thing which should be explored and implemented. There are several other initializers which can benefit from this optimization. (almost all initializers derived from TelemetryInitializerBase).

However, I do not think it's blocking this PR. It should be pursued as a separate improvement to benefit all initializers, including this.

Reply:
Reality is that we introduce yet another contention (because perf is not good already) that we should fix separately, but we likely never will. At least this work is not planned.  It seems doing a more performant thing is not that complicated, so why delay it for unknown amount of time?

Reply:
I'd prefer to not delay this fix for another milestone (after being pushed several times already). This is a constant source of support tickets with no end in sight.

Please post a new issue in github about the proposed way to avoid locking, I will take it for all the initializers. 

If perf becomes a concern due to this particular initializer, it can removed and replaced with the old initializer. (by user)